### PR TITLE
[CN-799]: Enhanced Jet Engine configuration

### PR DESCRIPTION
**Pull request by kutluhanmetin**
## Description

Enables `instance` and `edgeDefaults` Jet Configurations.
https://docs.hazelcast.com/hazelcast/5.2/configuration/jet-configuration#list-of-configuration-options

## User Impact

User can configure `instance` and `edgeDefaults` Jet Configuration parameters. Example configuration:
```
apiVersion: hazelcast.com/v1alpha1
kind: Hazelcast
metadata:
  name: hazelcast
spec:
  clusterSize: 3
  repository: 'docker.io/hazelcast/hazelcast-enterprise'
  jet:
    enabled: true
    resourceUploadEnabled: false
    instance:
      cooperativeThreadCount: 4
      flowControlPeriodMS: 100
      backupCount: 1
      scaleUpDelayMS: 10000
      losslessRestartEnabled: false
      maxProcessorAccumulatedRecords: 1000000000
    edgeDefaults:
      queueSize: 1024
      packetSizeLimit: 16384
      receiveWindowMultiplier: 3
  version: '5.1.4'
  licenseKeySecret: hazelcast-license-key
```

### CODE REVIEW

File: api/v1alpha1/hazelcast_types.go [hunk 1]
```diff
@@ -173,6 +173,70 @@ type JetEngineConfiguration struct {
...
+
+	// The duration of the interval between flow-control packets.
+	// +kubebuilder:default:=100
+	// +optional
+	FlowControlPeriodMS int `json:"flowControlPeriodMS"`
+
+	// The number of synchronous backups to configure on the IMap that Jet needs internally to store job metadata and snapshots.
+	// +kubebuilder:default:=1
+	// +optional
+	BackupCount int `json:"backupCount"`
```
Comment (mtyazici) [inline api/v1alpha1/hazelcast_types.go:201]: Same as `ScaleUpDelayMS`

File: api/v1alpha1/hazelcast_types.go [hunk 2]
```diff
@@ -173,6 +173,70 @@ type JetEngineConfiguration struct {
...
+	// Jet Edge Defaults Configuration
+	// +kubebuilder:default:={}
+	// +optional
+	EdgeDefaults *JetEdgeDefaults `json:"edgeDefaults"`
+}
+
+type JetInstance struct {
+	// The number of threads Jet creates in its cooperative multithreading pool.
+	// +optional
+	CooperativeThreadCount int `json:"cooperativeThreadCount"`
```
Comment (mtyazici) [inline api/v1alpha1/hazelcast_types.go:191]: This one will act like its default value is 0, is this ok?

File: api/v1alpha1/hazelcast_types.go [hunk 3]
```diff
@@ -173,6 +173,70 @@ type JetEngineConfiguration struct {
...
+	// +optional
+	MaxProcessorAccumulatedRecords int `json:"maxProcessorAccumulatedRecords"`
+}
+
+// Returns true if Jet Instance section is configured.
+func (j *JetInstance) IsConfigured() bool {
+	return j != nil && !(*j == (JetInstance{}))
+}
+
+type JetEdgeDefaults struct {
```
Comment (mtyazici) [inline api/v1alpha1/hazelcast_types.go:223]: All these fields will have 0 values, is this ok?

**Review by mtyazici — COMMENTED** [review]
I left a few comments about newly exposed API fields.

**Review by mtyazici — APPROVED** [review]
LGTM 👍🏼

File: api/v1alpha1/hazelcast_types.go [hunk 4]
```diff
@@ -173,6 +173,70 @@ type JetEngineConfiguration struct {
...
+
+	// The number of synchronous backups to configure on the IMap that Jet needs internally to store job metadata and snapshots.
+	// +kubebuilder:default:=1
+	// +optional
+	BackupCount int `json:"backupCount,omitempty"`
+
+	// The delay after which the auto-scaled jobs restart if a new member joins the cluster.
+	// +kubebuilder:default:=10000
+	// +optional
+	ScaleUpDelayMS int `json:"scaleUpDelayMS,omitempty"`
```
[the remark this answers is not in the document]
Comment (kutluhanmetin) [reply api/v1alpha1/hazelcast_types.go]: Done

File: api/v1alpha1/hazelcast_validation.go [hunk 5]
```diff
@@ -441,3 +445,28 @@ func ValidateNotUpdatableHzPersistenceFields(current, last *HazelcastPersistence
...
+	j := h.Spec.JetEngineConfiguration
+	p := h.Spec.Persistence
+
+	if !j.IsConfigured() {
+		return nil
+	}
+	if !j.Instance.IsConfigured() {
+		return nil
+	}
+	if j.Instance.BackupCount > 6 {
```
Comment (SeriyBg) [inline api/v1alpha1/hazelcast_validation.go:461]:
```suggestion
	if j.Instance.BackupCount > n.MaxBackupCount {
```

File: api/v1alpha1/hazelcast_types.go [hunk 6]
```diff
@@ -204,6 +204,70 @@ type JetEngineConfiguration struct {
...
+	// +optional
+	CooperativeThreadCount int32 `json:"cooperativeThreadCount,omitempty"`
+
+	// The duration of the interval between flow-control packets.
+	// +kubebuilder:default:=100
+	// +optional
+	FlowControlPeriodMillis int32 `json:"flowControlPeriodMillis,omitempty"`
+
+	// The number of synchronous backups to configure on the IMap that Jet needs internally to store job metadata and snapshots.
+	// +kubebuilder:default:=1
```
Comment (SeriyBg) [inline api/v1alpha1/hazelcast_types.go:230]: How about adding `// +kubebuilder:validation:Maximum:=6`?

**Review by SeriyBg — APPROVED** [review]

**Pull request merged**